### PR TITLE
Support WS Ping-Pong for Basic Station

### LINKED ISSUE #2003: Basic Station WS Ping-Pong
#### Summary

Implement WS Ping for Basic Station frontend.

#### Why do we need this?

To prevent our proxy from dropping WS connections due to inactivity of the gateway.
Refs: #1729 

#### What is already there? What do you see now?

No Ping sent.

#### What is missing? What do you want to see?

Send WS pings periodically (configurable).

#### Environment

`ttn-lw-stack v3.5.2`

#### How do you propose to implement this?

Add periodic pings to BS frontend.

#### Can you do this yourself and submit a Pull Request?

Yes

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #2003 

#### Changes

- Make Server ping interval configrable
- Add ping-pong handlers
- Tests

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 1]
```diff
@@ -63,13 +63,14 @@ type srv struct {
...
 	useTrafficTLSAddress bool
+	wsPingInterval       time.Duration
 }
 
 func (*srv) Protocol() string            { return "basicstation" }
 func (*srv) SupportsDownlinkClaim() bool { return false }
 
 // New creates the Basic Station front end.
-func New(ctx context.Context, server io.Server, useTrafficTLSAddress bool) *echo.Echo {
+func New(ctx context.Context, server io.Server, useTrafficTLSAddress bool, wsPingInterval time.Duration) *echo.Echo {
```
Comment (htdvisser) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: If we keep adding config/options here, we may want to consider switching to variadic `Option`s arguments.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 2]
```diff
@@ -253,12 +255,39 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 	fps := conn.FrequencyPlans()
 	bandID := conn.BandID()
 
+	pingTicker := time.NewTicker(s.wsPingInterval)
+	defer pingTicker.Stop()
+
+	ws.SetPingHandler(func(data string) error {
+		logger := logger
```
Comment (htdvisser) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: This is not necessary, we're not mutating `logger` in here.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 3]
```diff
@@ -253,12 +255,39 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
...
+		if err := ws.WriteMessage(websocket.PongMessage, nil); err != nil {
+			logger.WithError(err).Warn("Failed to send pong")
+			return err
+		}
+		return nil
+	})
+
+	// Not all gateways support pongs to the server's pings.
+	ws.SetPongHandler(func(data string) error {
+		logger := logger
```
Comment (htdvisser) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: This is not necessary, we're not mutating `logger` in here.

File: pkg/gatewayserver/gatewayserver_test.go [hunk 4]
```diff
@@ -60,7 +60,8 @@ var (
 	unregisteredGatewayID  = "eui-bbff000000000000"
 	unregisteredGatewayEUI = types.EUI64{0xBB, 0xFF, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	timeout = (1 << 5) * test.Delay
+	timeout        = (1 << 5) * test.Delay
+	WSPingInterval = (1 << 3) * test.Delay
```
Comment (johanstokking) [inline pkg/gatewayserver/gatewayserver_test.go]: nit; `wsPingInterval`

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 1 above)
Comment (johanstokking) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Then we also need to have defaults if we don't specify those options, while we already have defaults for the component config.

I'd work with an options struct instead to group settings, if the goal is to reduce the size of the argument list.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 5]
```diff
@@ -253,12 +255,39 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 	fps := conn.FrequencyPlans()
 	bandID := conn.BandID()
 
+	pingTicker := time.NewTicker(s.wsPingInterval)
+	defer pingTicker.Stop()
+
+	ws.SetPingHandler(func(data string) error {
+		logger := logger
+		logger.Debug("Received ws ping from gateway, sending pong")
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: - `WebSocket`, or remove `ws`
- Actions in imperative tense; `send pong`

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 6]
```diff
@@ -253,12 +255,39 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
...
+			logger.WithError(err).Warn("Failed to send pong")
+			return err
+		}
+		return nil
+	})
+
+	// Not all gateways support pongs to the server's pings.
+	ws.SetPongHandler(func(data string) error {
+		logger := logger
+		logger.Debug("Received Pong from Gateway")
```
Comment (johanstokking) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: `pong` and `gateway`, harmonize with message above

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 1 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Yup I plan to do this in the future if we add need to add another field.

**Pull request merged**